### ISSUE TITLE
logger no longer works from the command line with syslog-ng over UNIX socket

### ISSUE THREAD

**Issue by zachl123** [karelzak/util-linux#225]
When logging to syslog-ng over a UNIX socket, logger used to replace the \n at the end of each line with \0 (rather than just stripping the \n)
you can see this in logger.c (2.22) where it says:
if (write(fd, buf, strlen(buf)+1) < 0)
Note the "+1" on the strlen which included the null terminator.

you can verify this issue by running:
printf "XYZZY\nABCDEFG\n" |/home/zlevow/dev/util-linux_2.26/logger

if logger talks over a linux socket, you will see the following in the log files:
XYZZY<13>ABCDEFG\n

**Comment by zachl123** [comment]
Proposed patch:
diff --git a/misc-utils/logger.c b/misc-utils/logger.c
index 8908dfc..df73dcc 100644
--- a/misc-utils/logger.c
+++ b/misc-utils/logger.c
@@ -376,7 +376,7 @@ static const char *rfc3164_current_time(void)
  * full blown RFC5425 (TLS) looks like it is too much for the
  * logger utility.
  */
-static void write_output(const struct logger_ctl *ctl, const char *const msg)
+static void write_output(const struct logger_ctl *ctl, const char *const msg, int unix_socket_append_null)
 {
        char *buf;
        const size_t len = xasprintf(&buf, "%s%s", ctl->hdr, msg);
@@ -394,6 +394,13 @@ static void write_output(const struct logger_ctl *ctl, const char *const msg)
                        if (write_all(ctl->fd, "\n", 1) < 0)
                                warn(_("write failed"));
                }
+               else if (ctl->socket_type != TYPE_UDP) {
+                       /* for a unix socket, we need to append null to then end
+                        * assuming not true of udp
+                        */
+                       if (unix_socket_append_null && write_all(ctl->fd, "\0", 1) < 0)
+                               warn(_("write failed"));
+               }
        }
        if (ctl->stderr_printout)
                fprintf(stderr, "%s\n", buf);
@@ -610,12 +617,12 @@ static void logger_command_line(const struct logger_ctl *ctl, char **argv)
        while (*argv) {
                len = strlen(*argv);
                if (endp < p + len && p != buf) {
-                       write_output(ctl, buf);
+                       write_output(ctl, buf, 0);
                        p = buf;
                }
                if (ctl->max_message_size < len) {
                        (*argv)[ctl->max_message_size] = '\0'; /* truncate */
-                       write_output(ctl, *argv++);
+                       write_output(ctl, *argv++, 0);
                        continue;
                }
                if (p != buf)
@@ -624,7 +631,7 @@ static void logger_command_line(const struct logger_ctl *ctl, char **argv)
                *(p += len) = '\0';
        }
        if (p != buf)
-               write_output(ctl, buf);
+               write_output(ctl, buf, 0);
        free(buf);
 }

@@ -676,7 +683,7 @@ static void logger_stdin(struct logger_ctl *ctl)
                buf[i] = '\0';

                if (i > 0 || !ctl->skip_empty_lines)
-                       write_output(ctl, buf);
+                       write_output(ctl, buf, 1);

                if (c == '\n')  /* discard line terminator */
                        c = getchar();

**Comment by zachl123** [comment]
```c
diff --git a/misc-utils/logger.c b/misc-utils/logger.c
index 8908dfc..df73dcc 100644
--- a/misc-utils/logger.c
+++ b/misc-utils/logger.c
@@ -376,7 +376,7 @@ static const char *rfc3164_current_time(void)
  * full blown RFC5425 (TLS) looks like it is too much for the
  * logger utility.
  */
-static void write_output(const struct logger_ctl *ctl, const char *const msg)
+static void write_output(const struct logger_ctl *ctl, const char *const msg, int unix_socket_append_null)
 {
        char *buf;
        const size_t len = xasprintf(&buf, "%s%s", ctl->hdr, msg);
@@ -394,6 +394,13 @@ static void write_output(const struct logger_ctl *ctl, const char *const msg)
                        if (write_all(ctl->fd, "\n", 1) < 0)
                                warn(_("write failed"));
                }
+               else if (ctl->socket_type != TYPE_UDP) {
+                       /* for a unix socket, we need to append null to then end
+                        * assuming not true of udp
+                        */
+                       if (unix_socket_append_null && write_all(ctl->fd, "\0", 1) < 0)
+                               warn(_("write failed"));
+               }
        }
        if (ctl->stderr_printout)
                fprintf(stderr, "%s\n", buf);
@@ -610,12 +617,12 @@ static void logger_command_line(const struct logger_ctl *ctl, char **argv)
        while (*argv) {
                len = strlen(*argv);
                if (endp < p + len && p != buf) {
-                       write_output(ctl, buf);
+                       write_output(ctl, buf, 0);
                        p = buf;
                }
                if (ctl->max_message_size < len) {
                        (*argv)[ctl->max_message_size] = '\0'; /* truncate */
-                       write_output(ctl, *argv++);
+                       write_output(ctl, *argv++, 0);
                        continue;
                }
                if (p != buf)
@@ -624,7 +631,7 @@ static void logger_command_line(const struct logger_ctl *ctl, char **argv)
                *(p += len) = '\0';
        }
        if (p != buf)
-               write_output(ctl, buf);
+               write_output(ctl, buf, 0);
        free(buf);
 }

@@ -676,7 +683,7 @@ static void logger_stdin(struct logger_ctl *ctl)
                buf[i] = '\0';

                if (i > 0 || !ctl->skip_empty_lines)
-                       write_output(ctl, buf);
+                       write_output(ctl, buf, 1);

                if (c == '\n')  /* discard line terminator */
                        c = getchar();

```

**Comment by karelzak** [comment]
Thanks for your report.

I have applied a different patch. You're right that the original logger writes extra \0, but it seems that extra \n used by the new implementation is good enough (tested with syslog-ng-3.6.2). 

The problem is that write_output() add the extra \n only when ctl->socket_type is TYPE_TPC. Unfortunately, the connection initialization functions do not set the proper TYPE_* to the control struct and there is the default ALL_TYPES. This is reason why the problem disappears when you specify --tcp or --udp on logger command line.

I have fixed the connection functions, so now write_output() has all necessary information about the connection type.
